File: backend/app/schemas/production_advanced.py
"""
Pydantic schemas for the Advanced Production Module.
Closely mirrors the user-provided CSV column structures.
"""
from __future__ import annotations
from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field, model_validator


# ── Work Center ────────────────────────────────────────────────────────────────

class WorkCenterCreate(BaseModel):
    work_center_id: str = Field(..., max_length=50)
    name:           str = Field(..., max_length=255)
    type:           str = Field("MACHINE")
    capacity:       Optional[Decimal] = None
    capacity_uom:   Optional[str]     = None
    location:       Optional[str]     = None
    department:     Optional[str]     = None
    setup_time_min: Optional[int]     = None
    ideal_cycle_time_sec: Optional[Decimal] = None
    status:         str = Field("active")
    notes:          Optional[str]     = None

class WorkCenterUpdate(BaseModel):
    name:           Optional[str]     = None
    type:           Optional[str]     = None
    capacity:       Optional[Decimal] = None
    capacity_uom:   Optional[str]     = None
    location:       Optional[str]     = None
    department:     Optional[str]     = None
    setup_time_min: Optional[int]     = None
    ideal_cycle_time_sec: Optional[Decimal] = None
    status:         Optional[str]     = None
    notes:          Optional[str]     = None

class WorkCenterRead(BaseModel):
    id:             UUID
    work_center_id: str
    name:           str
    type:           str
    capacity:       Optional[Decimal]
    capacity_uom:   Optional[str]
    location:       Optional[str]
    department:     Optional[str]
    setup_time_min: Optional[int]
    status:         str
    notes:          Optional[str]
    created_at:     datetime

    model_config = {"from_attributes": True}


# ── Routing ────────────────────────────────────────────────────────────────────

class RoutingStepCreate(BaseModel):
    step_number:           int
    operation:             str = Field(..., max_length=255)
    work_center_id:        UUID
    standard_time_minutes: int   = Field(0)
    setup_time_minutes:    Optional[int] = None
    is_parallel:           bool  = False
    notes:                 Optional[str] = None

class RoutingCreate(BaseModel):
    routing_id: str = Field(..., max_length=50)
    product_id: UUID
    version:    int  = 1
    name:       Optional[str] = None
    is_active:  bool = True
    notes:      Optional[str] = None

class RoutingUpdate(BaseModel):
    name:      Optional[str]  = None
    is_active: Optional[bool] = None
    notes:     Optional[str]  = None

class RoutingStepRead(BaseModel):
    id:                    UUID
    routing_id:            UUID
    step_number:           int
    operation:             str
    work_center_id:        UUID
    work_center_name:      Optional[str] = None
    standard_time_minutes: int
    setup_time_minutes:    Optional[int]
    is_parallel:           bool

    model_config = {"from_attributes": True}

class RoutingRead(BaseModel):
    id:         UUID
    routing_id: str
    product_id: UUID
    product_name: Optional[str] = None
    product_sku:  Optional[str] = None
    version:    int
    name:       Optional[str]
    is_active:  bool
    steps:      List[RoutingStepRead] = []

    model_config = {"from_attributes": True}


# ── Work Orders ────────────────────────────────────────────────────────────────

class WorkOrderCreate(BaseModel):
    work_order_id:       str  = Field(..., max_length=50)
    production_order_id: UUID
    work_center_id:      UUID
    operation:           str  = Field(..., max_length=255)
    operator:            Optional[str] = None
    routing_step_id:     Optional[UUID] = None
    status:              str  = "planned"
    planned_quantity:    Optional[Decimal] = None
    start_time:          Optional[datetime] = None
    end_time:            Optional[datetime] = None
    planned_start:       Optional[datetime] = None
    planned_end:         Optional[datetime] = None
    notes:               Optional[str] = None

class WorkOrderUpdate(BaseModel):
    status:           Optional[str]      = None
    operator:         Optional[str]      = None
    actual_quantity:  Optional[Decimal]  = None
    start_time:       Optional[datetime] = None
    end_time:         Optional[datetime] = None
    notes:            Optional[str]      = None

class WorkOrderRead(BaseModel):
    id:                  UUID
    work_order_id:       str
    production_order_id: UUID
    order_no:            Optional[str]  = None
    work_center_id:      UUID
    work_center_name:    Optional[str] = None
    operation:           str
    operator:            Optional[str]
    status:              str
    planned_quantity:    Optional[Decimal]
    actual_quantity:     Optional[Decimal]
    start_time:          Optional[datetime]
    end_time:            Optional[datetime]
    planned_start:       Optional[datetime]
    planned_end:         Optional[datetime]
    created_at:          datetime

    model_config = {"from_attributes": True}


# ── Shifts ─────────────────────────────────────────────────────────────────────

class ShiftCreate(BaseModel):
    shift_code:     str = Field(..., max_length=50)
    name:           str = Field(..., max_length=255)
    start_time_str: str = Field(..., description="HH:MM format")
    end_time_str:   str = Field(..., description="HH:MM format")
    duration_hours: Decimal
    is_active:      bool = True
    notes:          Optional[str] = None

class ShiftUpdate(BaseModel):
    name:           Optional[str]  = None
    start_time_str: Optional[str]  = None
    end_time_str:   Optional[str]  = None
    duration_hours: Optional[Decimal] = None
    is_active:      Optional[bool] = None

class ShiftRead(BaseModel):
    id:             UUID
    shift_code:     str
    name:           str
    start_time_str: str
    end_time_str:   str
    duration_hours: Decimal
    is_active:      bool

    model_config = {"from_attributes": True}


# ── Production Schedule ────────────────────────────────────────────────────────

class ProductionScheduleCreate(BaseModel):
    schedule_id:         str = Field(..., max_length=50)
    production_order_id: UUID
    work_center_id:      UUID
    shift_id:            Optional[UUID]   = None
    shift_name:          Optional[str]    = None
    start_time:          datetime
    end_time:            datetime
    status:              str = "planned"
    priority:            str = "medium"
    notes:               Optional[str]    = None

class ProductionScheduleUpdate(BaseModel):
    status:     Optional[str]      = None
    priority:   Optional[str]      = None
    start_time: Optional[datetime] = None
    end_time:   Optional[datetime] = None
    notes:      Optional[str]      = None

class ProductionScheduleRead(BaseModel):
    id:                  UUID
    schedule_id:         str
    production_order_id: UUID
    order_no:            Optional[str] = None
    work_center_id:      UUID
    work_center_name:    Optional[str] = None
    shift_id:            Optional[UUID]
    shift_name:          Optional[str]
    start_time:          datetime
    end_time:            datetime
    status:              str
    priority:            str
    conflict_flag:       bool

    model_config = {"from_attributes": True}


# ── Time Tracking ──────────────────────────────────────────────────────────────

class TimeTrackingCreate(BaseModel):
    log_id:               str  = Field(..., max_length=50)
    work_order_id:        UUID
    actual_start:         datetime
    actual_end:           Optional[datetime] = None
    duration_planned_min: Optional[int]      = None
    downtime_minutes:     Optional[int]      = 0
    category:             str = "productive"
    reason:               Optional[str]      = None
    notes:                Optional[str]      = None

class TimeTrackingUpdate(BaseModel):
    actual_end:       Optional[datetime] = None
    downtime_minutes: Optional[int]      = None
    category:         Optional[str]      = None
    reason:           Optional[str]      = None

class TimeTrackingRead(BaseModel):
    id:                   UUID
    log_id:               str
    work_order_id:        UUID
    work_order_code:      Optional[str] = None
    actual_start:         datetime
    actual_end:           Optional[datetime]
    duration_actual_min:  Optional[int]
    duration_planned_min: Optional[int]
    downtime_minutes:     Optional[int]
    category:             str
    reason:               Optional[str]
    efficiency_pct:       Optional[Decimal]

    model_config = {"from_attributes": True}


# ── Downtime Events ────────────────────────────────────────────────────────────

class DowntimeEventCreate(BaseModel):
    downtime_id:    str  = Field(..., max_length=50)
    work_center_id: UUID
    machine_id:     Optional[str]      = None
    work_order_id:  Optional[UUID]     = None
    start_time:     datetime
    end_time:       Optional[datetime] = None
    reason:         str
    category:       str = "unplanned"
    reported_by:    Optional[str]      = None
    impact_units:   Optional[int]      = None
    notes:          Optional[str]      = None

class DowntimeEventUpdate(BaseModel):
    end_time:     Optional[datetime] = None
    reason:       Optional[str]      = None
    category:     Optional[str]      = None
    impact_units: Optional[int]      = None
    notes:        Optional[str]      = None

class DowntimeEventRead(BaseModel):
    id:              UUID
    downtime_id:     str
    work_center_id:  UUID
    work_center_name: Optional[str] = None
    machine_id:      Optional[str]
    work_order_id:   Optional[UUID]
    start_time:      datetime
    end_time:        Optional[datetime]
    duration_min:    Optional[int]
    reason:          str
    category:        str
    reported_by:     Optional[str]

    model_config = {"from_attributes": True}


# ── QC Inspections ─────────────────────────────────────────────────────────────

class AdvQCResultCreate(BaseModel):
    test_type:     str
    actual_value:  Optional[Decimal] = None
    min_value:     Optional[Decimal] = None
    max_value:     Optional[Decimal] = None
    unit:          Optional[str]     = None
    result:        str = "pass"
    notes:         Optional[str]     = None

class AdvQCInspectionCreate(BaseModel):
    qc_id:               str  = Field(..., max_length=50)
    production_order_id: UUID
    work_order_id:       Optional[UUID]     = None
    inspection_type:     str  = "inline"
    status:              str  = "pending"
    checked_by:          Optional[str]      = None
    inspected_at:        Optional[datetime] = None
    batch_ref:           Optional[str]      = None
    notes:               Optional[str]      = None
    results:             List[AdvQCResultCreate] = []

class AdvQCInspectionUpdate(BaseModel):
    status:      Optional[str]      = None
    checked_by:  Optional[str]      = None
    inspected_at: Optional[datetime] = None
    notes:        Optional[str]     = None

class AdvQCResultRead(BaseModel):
    id:            UUID
    inspection_id: UUID
    test_type:     str
    actual_value:  Optional[Decimal]
    min_value:     Optional[Decimal]
    max_value:     Optional[Decimal]
    unit:          Optional[str]
    result:        str

    model_config = {"from_attributes": True}

class AdvQCInspectionRead(BaseModel):
    id:                  UUID
    qc_id:               str
    production_order_id: UUID
    order_no:            Optional[str] = None
    work_order_id:       Optional[UUID]
    inspection_type:     str
    status:              str
    checked_by:          Optional[str]
    inspected_at:        Optional[datetime]
    batch_ref:           Optional[str]
    results:             List[AdvQCResultRead] = []

    model_config = {"from_attributes": True}


# ── Waste & Yield ──────────────────────────────────────────────────────────────

class WasteRecordCreate(BaseModel):
    waste_id:            str  = Field(..., max_length=50)
    production_order_id: UUID
    work_order_id:       Optional[UUID]    = None
    material_id:         Optional[UUID]    = None
    material_code:       Optional[str]     = None
    waste_type:          str  = "MATERIAL"
    waste_category:      str  = "normal"
    expected_qty:        Decimal
    actual_qty:          Decimal
    loss_qty:            Optional[Decimal] = None   # auto = expected - actual
    uom:                 str  = "KG"
    reason:              Optional[str]     = None
    notes:               Optional[str]     = None

    @model_validator(mode="after")
    def compute_loss(self) -> "WasteRecordCreate":
        if self.loss_qty is None:
            self.loss_qty = self.expected_qty - self.actual_qty
        return self

class WasteRecordUpdate(BaseModel):
    actual_qty:     Optional[Decimal] = None
    loss_qty:       Optional[Decimal] = None
    waste_category: Optional[str]     = None
    reason:         Optional[str]     = None

class WasteRecordRead(BaseModel):
    id:                  UUID
    waste_id:            str
    production_order_id: UUID
    order_no:            Optional[str]  = None
    material_code:       Optional[str]
    waste_type:          str
    waste_category:      str
    expected_qty:        Decimal
    actual_qty:          Decimal
    loss_qty:            Decimal
    loss_pct:            Optional[Decimal]
    uom:                 Optional[str]
    reason:              Optional[str]
    is_anomaly:          bool

    model_config = {"from_attributes": True}


# ── Batch / Lot Tracking ───────────────────────────────────────────────────────

class BatchLotCreate(BaseModel):
    batch_id:            str  = Field(..., max_length=100)
    production_order_id: UUID
    product_id:          UUID
    product_sku:         Optional[str]  = None
    quantity:            Decimal
    uom:                 str  = "KG"
    manufacture_date:    Optional[date] = None
    expiry_date:         Optional[date] = None
    status:              str  = "quarantine"
    traceability_code:   Optional[str]  = None
    warehouse_id:        Optional[UUID] = None
    notes:               Optional[str]  = None

class BatchLotUpdate(BaseModel):
    status:             Optional[str]  = None
    released_by:        Optional[str]  = None
    quarantine_reason:  Optional[str]  = None
    warehouse_id:       Optional[UUID] = None
    notes:              Optional[str]  = None

class BatchLotRead(BaseModel):
    id:                  UUID
    batch_id:            str
    production_order_id: UUID
    order_no:            Optional[str]  = None
    product_id:          UUID
    product_name:        Optional[str]  = None
    product_sku:         Optional[str]
    quantity:            Decimal
    uom:                 str
    manufacture_date:    Optional[date]
    expiry_date:         Optional[date]
    status:              str
    traceability_code:   Optional[str]
    warehouse_id:        Optional[UUID]

    model_config = {"from_attributes": True}


# ── Labor Tracking ─────────────────────────────────────────────────────────────

class LaborLogCreate(BaseModel):
    labor_id:       str  = Field(..., max_length=50)
    work_order_id:  UUID
    employee_id:    Optional[UUID]     = None
    employee_code:  Optional[str]      = None
    operator_name:  Optional[str]      = None
    start_time:     Optional[datetime] = None
    end_time:       Optional[datetime] = None
    hours_worked:   Optional[Decimal]  = None
    role:           Optional[str]      = None
    activity_type:  str  = "production"
    pieces_produced: Optional[int]     = None
    notes:          Optional[str]      = None

class LaborLogUpdate(BaseModel):
    end_time:       Optional[datetime] = None
    hours_worked:   Optional[Decimal]  = None
    pieces_produced: Optional[int]     = None
    notes:          Optional[str]      = None

class LaborLogRead(BaseModel):
    id:             UUID
    labor_id:       str
    work_order_id:  UUID
    work_order_code: Optional[str] = None
    employee_id:    Optional[UUID]
    employee_code:  Optional[str]
    operator_name:  Optional[str]
    hours_worked:   Optional[Decimal]
    role:           Optional[str]
    activity_type:  str
    pieces_produced: Optional[int]
    created_at:     datetime

    model_config = {"from_attributes": True}


# ── OEE Records ────────────────────────────────────────────────────────────────

class OEERecordCreate(BaseModel):
    oee_id:                     str  = Field(..., max_length=50)
    work_center_id:             UUID
    machine_id:                 Optional[str]  = None
    shift_id:                   Optional[UUID] = None
    record_date:                date
    planned_production_time_min: int
    actual_production_time_min:  Optional[int]     = None
    downtime_min:               Optional[int]      = 0
    ideal_cycle_time_sec:       Optional[Decimal]  = None
    total_units_produced:       Optional[int]      = 0
    good_units_produced:        Optional[int]      = 0
    # Accept pre-computed KPIs from CSV
    availability:               Optional[Decimal]  = None
    performance:                Optional[Decimal]  = None
    quality:                    Optional[Decimal]  = None
    oee_score:                  Optional[Decimal]  = None
    notes:                      Optional[str]      = None

class OEERecordUpdate(BaseModel):
    actual_production_time_min: Optional[int]     = None
    downtime_min:               Optional[int]      = None
    total_units_produced:       Optional[int]      = None
    good_units_produced:        Optional[int]      = None
    notes:                      Optional[str]      = None

class OEERecordRead(BaseModel):
    id:                         UUID
    oee_id:                     str
    work_center_id:             UUID
    work_center_name:           Optional[str] = None
    machine_id:                 Optional[str]
    shift_id:                   Optional[UUID]
    shift_name:                 Optional[str] = None
    record_date:                date
    planned_production_time_min: int
    actual_production_time_min:  Optional[int]
    downtime_min:               Optional[int]
    total_units_produced:       Optional[int]
    good_units_produced:        Optional[int]
    availability:               Optional[Decimal]
    performance:                Optional[Decimal]
    quality:                    Optional[Decimal]
    oee_score:                  Optional[Decimal]
    is_low_oee:                 bool
    is_high_downtime:           bool

    model_config = {"from_attributes": True}


# ── KPI / Analytics Schemas ────────────────────────────────────────────────────

class OEESummary(BaseModel):
    work_center_id:   UUID
    work_center_name: Optional[str]
    period_from:      date
    period_to:        date
    avg_availability: Optional[Decimal]
    avg_performance:  Optional[Decimal]
    avg_quality:      Optional[Decimal]
    avg_oee:          Optional[Decimal]
    total_downtime_min: int
    records_count:    int

class WasteSummary(BaseModel):
    production_order_id: UUID
    order_no:            Optional[str]
    total_expected:      Decimal
    total_actual:        Decimal
    total_loss:          Decimal
    avg_loss_pct:        Optional[Decimal]
    anomaly_count:       int

class ScheduleConflict(BaseModel):
    work_center_id:   UUID
    work_center_name: Optional[str]
    conflicting_schedules: List[str]
    overlap_minutes:  int

File: frontend/src/lib/productionAdvanced.ts
import { apiClient } from "./api";

const BASE = "/api/v1/production-adv";

// ── Types ───────────────────────────────────────────────────────────────────

export interface WorkCenter {
  id: string; work_center_id: string; name: string; type: string;
  capacity?: number; capacity_uom?: string; location?: string;
  department?: string; setup_time_min?: number; status: string; notes?: string;
}

export interface Routing {
  id: string; routing_id: string; product_id: string;
  product_name?: string; product_sku?: string;
  version: number; name?: string; is_active: boolean;
  steps: RoutingStep[];
}
export interface RoutingStep {
  id: string; routing_id: string; step_number: number; operation: string;
  work_center_id: string; work_center_name?: string;
  standard_time_minutes: number; setup_time_minutes?: number; is_parallel: boolean;
}

export interface WorkOrder {
  id: string; work_order_id: string; production_order_id: string;
  order_no?: string; work_center_id: string; work_center_name?: string;
  operation: string; operator?: string; status: string;
  planned_quantity?: number; actual_quantity?: number;
  start_time?: string; end_time?: string;
  planned_start?: string; planned_end?: string; created_at: string;
}

export interface Shift {
  id: string; shift_code: string; name: string;
  start_time_str: string; end_time_str: string;
  duration_hours: number; is_active: boolean;
}

export interface ProductionSchedule {
  id: string; schedule_id: string; production_order_id: string;
  order_no?: string; work_center_id: string; work_center_name?: string;
  shift_id?: string; shift_name?: string;
  start_time: string; end_time: string;
  status: string; priority: string; conflict_flag: boolean;
}

export interface TimeTracking {
  id: string; log_id: string; work_order_id: string; work_order_code?: string;
  actual_start: string; actual_end?: string;
  duration_actual_min?: number; duration_planned_min?: number;
  downtime_minutes?: number; category: string; reason?: string; efficiency_pct?: number;
}

export interface DowntimeEvent {
  id: string; downtime_id: string; work_center_id: string; work_center_name?: string;
  machine_id?: string; work_order_id?: string;
  start_time: string; end_time?: string; duration_min?: number;
  reason: string; category: string; reported_by?: string;
}

export interface QCInspection {
  id: string; qc_id: string; production_order_id: string; order_no?: string;
  work_order_id?: string; inspection_type: string; status: string;
  checked_by?: string; inspected_at?: string; batch_ref?: string;
  results: QCResult[];
}
export interface QCResult {
  id: string; inspection_id: string; test_type: string;
  actual_value?: number; min_value?: number; max_value?: number;
  unit?: string; result: string;
}

export interface WasteRecord {
  id: string; waste_id: string; production_order_id: string; order_no?: string;
  material_code?: string; waste_type: string; waste_category: string;
  expected_qty: number; actual_qty: number; loss_qty: number;
  loss_pct?: number; uom?: string; reason?: string; is_anomaly: boolean;
}

export interface BatchLot {
  id: string; batch_id: string; production_order_id: string; order_no?: string;
  product_id: string; product_name?: string; product_sku?: string;
  quantity: number; uom: string;
  manufacture_date?: string; expiry_date?: string; status: string;
  traceability_code?: string; warehouse_id?: string;
}

export interface LaborLog {
  id: string; labor_id: string; work_order_id: string; work_order_code?: string;
  employee_id?: string; employee_code?: string; operator_name?: string;
  hours_worked?: number; role?: string; activity_type: string;
  pieces_produced?: number; created_at: string;
}

export interface OEERecord {
  id: string; oee_id: string; work_center_id: string; work_center_name?: string;
  machine_id?: string; shift_id?: string; shift_name?: string;
  record_date: string; planned_production_time_min: number;
  actual_production_time_min?: number; downtime_min?: number;
  total_units_produced?: number; good_units_produced?: number;
  availability?: number; performance?: number; quality?: number; oee_score?: number;
  is_low_oee: boolean; is_high_downtime: boolean;
}

export interface OEESummaryRow {
  work_center_id: string; work_center_name?: string;
  avg_availability?: number; avg_performance?: number;
  avg_quality?: number; avg_oee?: number;
  total_downtime_min: number; records_count: number;
}

export interface AIInsight {
  type: string; severity: string;
  machine?: string; date?: string;
  oee_score?: number; downtime_min?: number;
  loss_pct?: number; waste_id?: string;
  recommendation: string;
}

// ── API ─────────────────────────────────────────────────────────────────────

export const advProductionApi = {
  // Work Centers
  async listWorkCenters(p?: { status?: string }): Promise<WorkCenter[]> {
    const r = await apiClient.get(`${BASE}/work-centers`, { params: p });
    return r.data;
  },
  async createWorkCenter(d: Partial<WorkCenter>): Promise<WorkCenter> {
    const r = await apiClient.post(`${BASE}/work-centers`, d);
    return r.data;
  },
  async updateWorkCenter(id: string, d: Partial<WorkCenter>): Promise<WorkCenter> {
    const r = await apiClient.patch(`${BASE}/work-centers/${id}`, d);
    return r.data;
  },
  async deleteWorkCenter(id: string): Promise<void> {
    await apiClient.delete(`${BASE}/work-centers/${id}`).then((r) => r.data);
  },

  // Routings
  async listRoutings(p?: { product_id?: string }): Promise<Routing[]> {
    const r = await apiClient.get(`${BASE}/routings`, { params: p });
    return r.data;
  },
  async createRouting(d: object): Promise<Routing> {
    const r = await apiClient.post(`${BASE}/routings`, d);
    return r.data;
  },
  async deleteRouting(id: string): Promise<void> {
    await apiClient.delete(`${BASE}/routings/${id}`).then((r) => r.data);
  },
  async addRoutingStep(routingId: string, d: object): Promise<RoutingStep> {
    const r = await apiClient.post(`${BASE}/routings/${routingId}/steps`, d);
    return r.data;
  },

  // Work Orders
  async listWorkOrders(p?: { production_order_id?: string; status?: string }): Promise<WorkOrder[]> {
    const r = await apiClient.get(`${BASE}/work-orders`, { params: p });
    return r.data;
  },
  async createWorkOrder(d: object): Promise<WorkOrder> {
    const r = await apiClient.post(`${BASE}/work-orders`, d);
    return r.data;
  },
  async updateWorkOrder(id: string, d: object): Promise<WorkOrder> {
    const r = await apiClient.patch(`${BASE}/work-orders/${id}`, d);
    return r.data;
  },
  async startWorkOrder(id: string): Promise<WorkOrder> {
    const r = await apiClient.post(`${BASE}/work-orders/${id}/start`);
    return r.data;
  },
  async completeWorkOrder(id: string, qty?: number): Promise<WorkOrder> {
    const r = await apiClient.post(`${BASE}/work-orders/${id}/complete`, null, { params: { actual_quantity: qty } });
    return r.data;
  },
  async deleteWorkOrder(id: string): Promise<void> {
    await apiClient.delete(`${BASE}/work-orders/${id}`).then((r) => r.data);
  },

  // Shifts
  async listShifts(): Promise<Shift[]> {
    const r = await apiClient.get(`${BASE}/shifts`);
    return r.data;
  },
  async createShift(d: object): Promise<Shift> {
    const r = await apiClient.post(`${BASE}/shifts`, d);
    return r.data;
  },
  async deleteShift(id: string): Promise<void> {
    await apiClient.delete(`${BASE}/shifts/${id}`).then((r) => r.data);
  },

  // Schedules
  async listSchedules(p?: { work_center_id?: string; status?: string; date_from?: string; date_to?: string }): Promise<ProductionSchedule[]> {
    const r = await apiClient.get(`${BASE}/schedules`, { params: p });
    return r.data;
  },
  async createSchedule(d: object): Promise<ProductionSchedule> {
    const r = await apiClient.post(`${BASE}/schedules`, d);
    return r.data;
  },
  async getConflicts(): Promise<ProductionSchedule[]> {
    const r = await apiClient.get(`${BASE}/schedules/conflicts`);
    return r.data;
  },
  async deleteSchedule(id: string): Promise<void> {
    await apiClient.delete(`${BASE}/schedules/${id}`).then((r) => r.data);
  },

  // Time Tracking
  async listTimeTracking(p?: { work_order_id?: string }): Promise<TimeTracking[]> {
    const r = await apiClient.get(`${BASE}/time-tracking`, { params: p });
    return r.data;
  },
  async createTimeTracking(d: object): Promise<TimeTracking> {
    const r = await apiClient.post(`${BASE}/time-tracking`, d);
    return r.data;
  },
  async deleteTimeTracking(id: string): Promise<void> {
    await apiClient.delete(`${BASE}/time-tracking/${id}`).then((r) => r.data);
  },

  // Downtime
  async listDowntime(p?: { work_center_id?: string; category?: string }): Promise<DowntimeEvent[]> {
    const r = await apiClient.get(`${BASE}/downtime`, { params: p });
    return r.data;
  },
  async createDowntime(d: object): Promise<DowntimeEvent> {
    const r = await apiClient.post(`${BASE}/downtime`, d);
    return r.data;
  },
  async updateDowntime(id: string, d: object): Promise<DowntimeEvent> {
    const r = await apiClient.patch(`${BASE}/downtime/${id}`, d);
    return r.data;
  },
  async deleteDowntime(id: string): Promise<void> {
    await apiClient.delete(`${BASE}/downtime/${id}`).then((r) => r.data);
  },

  // QC
  async listQC(p?: { production_order_id?: string; status?: string }): Promise<QCInspection[]> {
    const r = await apiClient.get(`${BASE}/qc`, { params: p });
    return r.data;
  },
  async createQC(d: object): Promise<QCInspection> {
    const r = await apiClient.post(`${BASE}/qc`, d);
    return r.data;
  },
  async updateQC(id: string, d: object): Promise<QCInspection> {
    const r = await apiClient.patch(`${BASE}/qc/${id}`, d);
    return r.data;
  },
  async deleteQC(id: string): Promise<void> {
    await apiClient.delete(`${BASE}/qc/${id}`).then((r) => r.data);
  },

  // Waste
  async listWaste(p?: { production_order_id?: string; is_anomaly?: boolean }): Promise<WasteRecord[]> {
    const r = await apiClient.get(`${BASE}/waste`, { params: p });
    return r.data;
  },
  async createWaste(d: object): Promise<WasteRecord> {
    const r = await apiClient.post(`${BASE}/waste`, d);
    return r.data;
  },
  async deleteWaste(id: string): Promise<void> {
    await apiClient.delete(`${BASE}/waste/${id}`).then((r) => r.data);
  },

  // Batch Lots
  async listBatchLots(p?: { production_order_id?: string; status?: string }): Promise<BatchLot[]> {
    const r = await apiClient.get(`${BASE}/batch-lots`, { params: p });
    return r.data;
  },
  async createBatchLot(d: object): Promise<BatchLot> {
    const r = await apiClient.post(`${BASE}/batch-lots`, d);
    return r.data;
  },
  async updateBatchLot(id: string, d: object): Promise<BatchLot> {
    const r = await apiClient.patch(`${BASE}/batch-lots/${id}`, d);
    return r.data;
  },
  async deleteBatchLot(id: string): Promise<void> {
    await apiClient.delete(`${BASE}/batch-lots/${id}`).then((r) => r.data);
  },

  // Labor
  async listLabor(p?: { work_order_id?: string }): Promise<LaborLog[]> {
    const r = await apiClient.get(`${BASE}/labor`, { params: p });
    return r.data;
  },
  async createLabor(d: object): Promise<LaborLog> {
    const r = await apiClient.post(`${BASE}/labor`, d);
    return r.data;
  },
  async deleteLabor(id: string): Promise<void> {
    await apiClient.delete(`${BASE}/labor/${id}`).then((r) => r.data);
  },

  // OEE
  async listOEE(p?: { work_center_id?: string; date_from?: string; date_to?: string }): Promise<OEERecord[]> {
    const r = await apiClient.get(`${BASE}/oee`, { params: p });
    return r.data;
  },
  async createOEE(d: object): Promise<OEERecord> {
    const r = await apiClient.post(`${BASE}/oee`, d);
    return r.data;
  },
  async updateOEE(id: string, d: object): Promise<OEERecord> {
    const r = await apiClient.patch(`${BASE}/oee/${id}`, d);
    return r.data;
  },
  async deleteOEE(id: string): Promise<void> {
    await apiClient.delete(`${BASE}/oee/${id}`).then((r) => r.data);
  },
  async oeeeSummary(p?: { work_center_id?: string; date_from?: string; date_to?: string }): Promise<OEESummaryRow[]> {
    const r = await apiClient.get(`${BASE}/oee/summary`, { params: p });
    return r.data;
  },

  // AI Insights
  async aiInsights(p?: { work_center_id?: string }): Promise<{ total_insights: number; insights: AIInsight[] }> {
    const r = await apiClient.get(`${BASE}/ai-insights`, { params: p });
    return r.data;
  },
};
